ddapm_test_agent/trace.py

```diff
@@ -124,6 +126,24 @@ def verify_span(d: Any) -> Span:
             for k, v in d["meta"].items():
                 assert isinstance(k, str), f"Expected key 'meta.{k}' to be of type: 'str', got: {type(k)}"
                 assert isinstance(v, str), f"Expected value of key 'meta.{k}' to be of type: 'str', got: {type(v)}"
+        if "meta_struct" in d:
+            assert isinstance(d["meta_struct"], dict)
+            for k, v in d["meta_struct"].items():
+                assert isinstance(k, str), f"Expected key 'meta_struct.{k}' to be of type: 'str', got: {type(k)}"
+                assert isinstance(
+                    v, bytes
+                ), f"Expected msgpack'd value of key 'meta_struct.{k}' to be of type: 'bytes', got: {type(v)}"
+            # Decode meta_struct msgpack values
+            decoded_meta_struct = {key: msgpack.unpackb(val_bytes) for key, val_bytes in d["meta_struct"].items()}
+            for k, val in decoded_meta_struct.items():
+                assert isinstance(
+                    val, dict
+                ), f"Expected msgpack decoded value of key 'meta_struct.{k}' to be of type: 'dict', got: {type(val)}"
+                for inner_k in val:
+                    assert isinstance(
+                        inner_k, str
+                    ), f"Expected key 'meta_struct.{k}.{inner_k}' to be of type: 'str', got: {type(inner_k)}"
+            d["meta_struct"] = decoded_meta_struct
```

Comment:
it's pretty unintuitive that verify_span actually mutates it. Would it be possible to do it in `_trace_decoder_flexible` instead?

Reply:
@Kyle-Verhoog  The code in `_trace_decoder_flexible` looks like it's done to unmarshal only json and not msgpack. So I don't know how I should do the same for msgpack (which is the one I am interested in). I know this is not ideal to put it in `verify_span` but I did not found a better place. I can try to create a new function and call it before `verify_span` if you want but :shrug:

Reply:
ah crap you're right. Yeah I think a new function that does it and added to the `decode_v0{4,5,7}` methods would be the way to go then. WDYT?

Reply:
ok let me do this then